module/zfs/zfs_vfsops.c

```diff
@@ -1769,6 +1769,13 @@ zfs_remount(struct super_block *sb, int *flags, zfs_mnt_t *zm)
 	if (error)
 		return (error);
 
+	if ((dmu_objset_is_snapshot(zfsvfs->z_os) ||
+	    !spa_writeable(dmu_objset_spa(zfsvfs->z_os))) &&
+	    !(*flags & MS_RDONLY)) {
+		*flags |= MS_RDONLY;
+		return (EROFS);
+	}
+
 	zfs_unregister_callbacks(zfsvfs);
```

Comment:
```
[16498.559788] VERIFY(list_is_empty(list)) failed
[16498.560525] PANIC at list.h:88:list_destroy()
[16498.561242] Showing stack for process 908
[16498.561967] CPU: 0 PID: 908 Comm: dbu_evict Tainted: P           OE   4.9.0 #4
[16498.564434] Hardware name: Bochs Bochs, BIOS Bochs 01/01/2011
[16498.566161]  0000000000000000 ffffffff88129055 ffffffffc0a4e081 ffffbf2040e73e10
[16498.569740]  ffffffffc07eb03f ffffa08c4a3586c0 ffffffff00000028 ffffbf2040e73e20
[16498.573670]  ffffbf2040e73dc0 6c28594649524556 655f73695f747369 73696c287974706d
[16498.575983] Call Trace:
[16498.576456]  [<ffffffff88129055>] ? dump_stack+0x5c/0x77
[16498.578307]  [<ffffffffc07eb03f>] ? spl_panic+0xbf/0xf0 [spl]
[16498.580339]  [<ffffffff883f8dce>] ? mutex_lock+0xe/0x30
[16498.582178]  [<ffffffff883f8dce>] ? mutex_lock+0xe/0x30
[16498.584078]  [<ffffffffc0986fb9>] ? refcount_remove_many+0x179/0x280 [zfs]
[16498.585849]  [<ffffffffc0930935>] ? dbuf_rele_and_unlock+0x1e5/0x450 [zfs]
[16498.586479]  [<ffffffffc0930b65>] ? dbuf_rele_and_unlock+0x415/0x450 [zfs]
[16498.588155]  [<ffffffffc098525a>] ? multilist_insert+0xca/0x1a0 [zfs]
[16498.590031]  [<ffffffffc091cd0a>] ? remove_reference+0x8a/0x180 [zfs]
[16498.591858]  [<ffffffff883f8dce>] ? mutex_lock+0xe/0x30
[16498.592541]  [<ffffffffc0986fb9>] ? refcount_remove_many+0x179/0x280 [zfs]
[16498.594420]  [<ffffffffc096bffe>] ? dsl_dir_evict+0x1ae/0x1e0 [zfs]
[16498.596174]  [<ffffffffc07e8eb9>] ? taskq_thread+0x269/0x520 [spl]
[16498.597929]  [<ffffffff87ea9170>] ? wake_up_q+0x60/0x60
[16498.599595]  [<ffffffffc07e8c50>] ? taskq_thread_spawn+0x50/0x50 [spl]
[16498.600174]  [<ffffffff87e9dc10>] ? kthread+0xe0/0x100
[16498.601646]  [<ffffffff87e2b76b>] ? __switch_to+0x2bb/0x700
[16498.602361]  [<ffffffff87e9db30>] ? kthread_park+0x60/0x60
[16498.604091]  [<ffffffff883fbaf5>] ? ret_from_fork+0x25/0x3
```

The issue here is that `zfs_unregister_callbacks()` does not execute `dsl_prop_unregister_all()` if `dmu_objset_is_snapshot(os)`, so we leave a dirtry/non-empty `ds->ds_prop_cbs` list when we remount successfully a snapshot (`-o remount,noxattr`).

Maybe we should skip callback registration in `zfs_remount()` when we are processing a snapshot.